### PR TITLE
[CPU] DetectionOutput migration on nGraph.

### PR DESCRIPTION
**Pull request by nshchego**
### Details:
 - *item1*
 - *...*

### Tickets:
 - *ticket-id*

### CODE REVIEW

File: inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp [hunk 1]
```diff
@@ -24,98 +29,85 @@ static bool SortScorePairDescend(const std::pair<float, T>& pair1,
...
 
-            with_add_box_pred = layer->insData.size() == 5;
-            _objectness_score = layer->GetParamAsFloat("objectness_score", 0.0f);
+            with_add_box_pred = op->get_input_size() == 5;
+            _objectness_score = attributes.objectness_score;
 
-            std::string code_type_str = layer->GetParamAsString("code_type", "caffe.PriorBoxParameter.CORNER");
-            _code_type = (code_type_str == "caffe.PriorBoxParameter.CENTER_SIZE" ? CodeType::CENTER_SIZE
-                                                                                 : CodeType::CORNER);
+            _code_type = (details::CaselessEq<std::string>()(attributes.code_type, "caffe.PriorBoxParameter.CENTER_SIZE") ?
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp:75]: We have to check that `_code_type` is from supported range inside `isSupportedOperation`.

File: inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp [hunk 2]
```diff
@@ -24,98 +29,85 @@ static bool SortScorePairDescend(const std::pair<float, T>& pair1,
...
-
-            InferenceEngine::SizeVector num_priors_actual_size{static_cast<size_t>(_num)};
-            _num_priors_actual = InferenceEngine::make_shared_blob<int>({Precision::I32, num_priors_actual_size, C});
-            _num_priors_actual->allocate();
-
-            std::vector<DataConfigurator> in_data_conf(layer->insData.size(), DataConfigurator(ConfLayout::PLN, Precision::FP32));
-            addConfig(layer, in_data_conf, {DataConfigurator(ConfLayout::PLN, Precision::FP32)});
+            _num = static_cast<int>(op->get_input_shape(idx_confidence)[0]);
+
+            _decoded_bboxes = std::shared_ptr<float>(new float[_num * _num_classes * _num_priors * 4], [](float *p) { delete[] p; });
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp:94]: lets use vectors

File: inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp [hunk 3]
```diff
@@ -24,98 +29,85 @@ static bool SortScorePairDescend(const std::pair<float, T>& pair1,
...
+            }
+        } catch (...) {
+            return false;
+        }
+        return true;
+    }
+
+    explicit DetectionOutputImpl(const std::shared_ptr<ngraph::Node>& op) {
         try {
-            if (layer->insData.size() != 3 && layer->insData.size() != 5)
```
[the remark this answers is not in the document]
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp:29]: Done

File: inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp (hunk 1 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp:88]: Done

File: inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp (hunk 2 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp]: Done

File: inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp [hunk 4]
```diff
@@ -24,98 +29,97 @@ static bool SortScorePairDescend(const std::pair<float, T>& pair1,
...
-            _decrease_label_id = layer->GetParamAsBool("decrease_label_id", false);
-            _normalized = layer->GetParamAsBool("normalized", true);
-            _image_height = layer->GetParamAsInt("input_height", 1);
-            _image_width = layer->GetParamAsInt("input_width", 1);
+            auto doOp = ngraph::as_type_ptr<const ngraph::op::v0::DetectionOutput>(op);
+            if (!doOp) {
+                errorMessage = "Node is not an instance of the DetectionOutput from the operations set v0.";
+                return false;
+            }
+            if (op->get_input_size() != 3 && op->get_input_size() != 5) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp:39]: As we discussed in one of the previous PRs - this shouldn't be part of `isSupportedOperation` function

File: inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp [hunk 5]
```diff
@@ -24,98 +29,97 @@ static bool SortScorePairDescend(const std::pair<float, T>& pair1,
...
+                errorMessage = "Invalid number of input edges.";
+                return false;
+            }
+            if (op->get_output_size() == 0) {
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp:43]: the same

File: inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp (hunk 4 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp]: FIxed

File: inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp (hunk 5 above)
Comment (nshchego) [reply inference-engine/src/mkldnn_plugin/nodes/detectionoutput.cpp]: Fixed

**Review by dmitry-gorokhov — APPROVED** [review]

**Pull request merged**